js_modules/dagster-ui/packages/ui-core/src/assets/catalog/useSelectionHealthData.tsx

```diff
@@ -0,0 +1,232 @@
+import React, {useContext, useLayoutEffect, useMemo, useReducer} from 'react';
+
+import {AssetHealthData, useAssetsHealthData} from '../../asset-data/AssetHealthDataProvider';
+import {AssetHealthFragment} from '../../asset-data/types/AssetHealthDataProvider.types';
+import {useAssetSelectionFiltering} from '../../asset-selection/useAssetSelectionFiltering';
+import {useUpdatingRef} from '../../hooks/useUpdatingRef';
+import {DeferredCallbackRegistry} from '../../util/DeferredCallbackRegistry';
+import {useAllAssets} from '../useAllAssets';
+
+type SelectionHealthData = {
+  liveDataByNode: Record<string, AssetHealthFragment>;
+  assetCount: number;
+  loading: boolean;
+};
+
+type SelectionFilterData = {
+  assets: ReturnType<typeof useAllAssets>['assets'];
+  loading: boolean;
+};
+
+export const SelectionHealthDataContext = React.createContext<{
+  watchSelection: (params: {
+    selection: string;
+    setHealthData?: (data: SelectionHealthData) => void;
+    setFilterData?: (data: SelectionFilterData) => void;
+  }) => () => void;
+}>({
+  watchSelection: () => () => {},
+});
+
+export const SelectionHealthDataProvider = ({children}: {children: React.ReactNode}) => {
+  const [selections, setSelections] = React.useState<Set<string>>(() => new Set());
+
+  /**
+   * Whenever new listeners are added to a registry, we need to force a re-render of the SelectionHealthDataObserver
+   * so that it picks up the listeners from the registry and uses that information to determine whether
+   * to fetch health data or calculate filter data.
+   * We do this instead of using state to avoid needing to manage references and create new objects.
+   */
+  const [, forceRerender] = useReducer((s: number) => s + 1, 0);
+
+  const registries: Record<string, SelectionRegistry> = useMemo(() => ({}), []);
```

Comment:
Don't think I've seen this approach before -- I feel like normally this would be in a ref.